cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/AbstractTransactionContext.java

```diff
@@ -0,0 +1,308 @@
+    checkForConflicts();
+    persist();
+    commit();
+
+    try {
+      postCommit();
+    } finally {
+      currentTx = null;
+      cleanup();
+    }
+  }
+
+  @Override
+  public void abort() throws TransactionFailureException {
+    abort(null);
+  }
+
+  @Override
+  public void abort(@Nullable TransactionFailureException cause) throws TransactionFailureException {
+    if (currentTx == null) {
+      // might be called by some generic exception handler even though already aborted/finished - we allow that
+      return;
+    }
+    try {
+      boolean success = true;
+      for (TransactionAware txAware : getTransactionAwares()) {
+        try {
+          success = txAware.rollbackTx() && success;
+        } catch (Throwable e) {
+          if (cause == null) {
+            cause = new TransactionFailureException(
+              String.format("Unable to roll back changes in transaction-aware '%s' for transaction %d. ",
+                            txAware.getTransactionAwareName(), currentTx.getTransactionId()), e);
+          } else {
+            cause.addSuppressed(e);
+          }
+          success = false;
+        }
+      }
+      if (success) {
+        txClient.abort(currentTx);
+      } else {
+        txClient.invalidate(currentTx.getTransactionId());
+      }
+      if (cause != null) {
+        throw cause;
+      }
+    } finally {
+      currentTx = null;
+      cleanup();
+    }
+  }
+
+  @Override
+  public void checkpoint() throws TransactionFailureException {
+    Preconditions.checkState(currentTx != null, "Cannot checkpoint tx that has not been started");
+    persist();
+    try {
+      currentTx = txClient.checkpoint(currentTx);
+      // update the current transaction with all TransactionAwares
+      for (TransactionAware txAware : getTransactionAwares()) {
+        txAware.updateTx(currentTx);
+      }
+    } catch (TransactionFailureException e) {
+      abort(e);
+    } catch (Throwable e) {
+      abort(new TransactionFailureException(
+        String.format("Exception from checkpoint for transaction %d.", currentTx.getTransactionId()), e));
+    }
+  }
+
+  @Nullable
+  @Override
+  public Transaction getCurrentTransaction() {
+    return currentTx;
+  }
+
+  /**
+   * Calls {@link TransactionAware#startTx(Transaction)} on all {@link TransactionAware}.
+   */
+  private void startAllTxAwares() throws TransactionFailureException {
+    for (TransactionAware txAware : getTransactionAwares()) {
+      try {
+        txAware.startTx(currentTx);
+      } catch (Throwable e) {
+        try {
+          txClient.abort(currentTx);
+          throw new TransactionFailureException(
+            String.format("Unable to start transaction-aware '%s' for transaction %d. ",
+                          txAware.getTransactionAwareName(), currentTx.getTransactionId()), e);
+        } finally {
+          currentTx = null;
+        }
```

Comment:
The current contract for the `cleanup()` is that it will only get called when the tx completed (either commit or aborted).